index.js

```diff
@@ -78,30 +85,38 @@ function wrapInControlFlow(flow, globalFn, fnName) {
 
         flow.execute(function controlFlowExecute() {
           return new webdriver.promise.Promise(function(fulfill, reject) {
+            function wrappedReject(err) {
+              var wrappedErr = new Error(err);
+              reject(wrappedErr);
+            }
             if (async) {
               // If testFn is async (it expects a done callback), resolve the promise of this
               // test whenever that callback says to.  Any promises returned from testFn are
               // ignored.
               var proxyDone = fulfill;
-              proxyDone.fail = function(err) {
-                var wrappedErr = new Error(err);
-                reject(wrappedErr);
-              };
+              proxyDone.fail = wrappedReject;
               testFn(proxyDone);
             } else {
               // Without a callback, testFn can return a promise, or it will
               // be assumed to have completed synchronously.
-              fulfill(testFn());
+              var ret = testFn();
+              if (webdriver.promise.isPromise(ret)) {
+                ret.then(fulfill, wrappedReject);
+              } else {
+                fulfill(ret);
+              }
             }
           }, flow);
-        }, 'Run ' + fnName + description + ' in control flow').then(seal(done), function(err) {
-          if (!err) {
-            err = new Error('Unknown Error');
-            err.stack = '';
+        }, 'Run ' + fnName + description + ' in control flow').then(
+          callWhenIdle.bind(null, flow, done), function(err) {
+            if (!err) {
+              err = new Error('Unknown Error');
+              err.stack = '';
+            }
+            err.stack = err.stack + '\nFrom asynchronous test: \n' + driverError.stack;
+            callWhenIdle(flow, done.fail.bind(done, err));
```

Comment:
We might want to call [`flow.reset()`](https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/lib/promise_exports_ControlFlow.html#reset) here instead of waiting for `idle`. What do you think?